mautrix_signal/portal.py

```diff
@@ -725,6 +725,52 @@ async def handle_matrix_leave(self, user: u.User) -> None:
                 await self.signal.leave_group(user.username, self.chat_id)
             # TODO cleanup if empty
 
+    async def kick_matrix(self, user: u.User | p.Puppet, source: u.User) -> None:
+        try:
+            await self.signal.update_group(
+                source.username, self.chat_id, remove_members=[user.address]
+            )
+        except Exception as e:
+            self.log.exception(f"Failed to kick Signal user: {e}")
+            info = await self.signal.get_group(source.username, self.chat_id)
+            if user.address in info.members:
+                await self.main_intent.invite_user(
+                    self.mxid,
+                    user.mxid,
+                    check_cache=True,
+                    reason=f"Failed to kick Signal user: {e}",
+                )
+                await user.intent_for(self).ensure_joined(self.mxid)
+
+    async def ban_matrix(self, user: u.User | p.Puppet, source: u.User) -> None:
+        try:
+            await self.signal.ban_user(source.username, self.chat_id, users=[user.address])
+        except Exception as e:
+            self.log.exception(f"Failed to ban Signal user: {e}")
+            info = await self.signal.get_group(source.username, self.chat_id)
+            if not info.banned_members or user.address not in info.banned_members:
+                await self.main_intent.unban_user(
+                    self.mxid, user.mxid, reason=f"Failed to ban Signal user: {e}"
+                )
+            if user.address in info.members:
+                await self.main_intent.invite_user(
+                    self.mxid,
+                    user.mxid,
+                    check_cache=True,
+                )
+                await user.intent_for(self).ensure_joined(self.mxid)
+
+    async def unban_matrix(self, user: u.User | p.Puppet, source: u.User) -> None:
+        try:
+            await self.signal.unban_user(source.username, self.chat_id, users=[user.address])
+        except Exception as e:
+            self.log.exception(f"Failed to unban Signal user: {e}")
+            info = await self.signal.get_group(source.username, self.chat_id)
+            if info.banned_members and user.address in info.banned_members:
```

Comment:
These `in banned_members` checks are still wrong

Reply:
OK, I did some more testing ((un)banning with insufficient rights, (un)banning when the user already is (un)banned). Everything works as expected.